### PR TITLE
[ONNX] Support optional type

### PR DESCRIPTION
**Pull request by garymm**
Based on #61938

### CODE REVIEW

**Comment by garymm** [comment]
@BowenBao can you review since you already started reviewing https://github.com/pytorch/pytorch/pull/61938?
If you're busy let me know and I'll ask someone else.

File: torch/csrc/jit/serialization/export.cpp [hunk 1]
```diff
@@ -1083,6 +1134,41 @@ void GraphEncoder::EncodeLocalFunction(
...
+    std::string name) {
+  if (node_type->kind() == TypeKind::TensorType) {
+    onnx::TypeProto_Tensor* tensor_type_proto =
+        type_proto->mutable_tensor_type();
+    TensorTypePtr t = node_type->cast<TensorType>();
+    if (t->dim()) {
+      onnx::TensorShapeProto* shape = tensor_type_proto->mutable_shape();
+      auto sizes = t->symbolic_sizes().sizes().value();
+      for (const auto i : c10::irange(sizes.size())) {
+        shape->add_dim();
```
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:1149]: nit: a little duplicated code with `EncodeValueInfoType` that can be extracted.

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -3398,7 +3398,8 @@ def Loop(ctx: torch.onnx.SymbolicContext, g, *inputs, **attrs):
             for i, b_in in enumerate(b.inputs()):
                 if i == 0 and i < len(inputs):
                     b_in.setType(inputs[i].type())
-                if i > 0 and (i + 1) < len(inputs):
+                # For optional block inputs, input type could be NoneType, which we don't want to propagate.
+                if i > 0 and (i + 1) < len(inputs) and not isinstance(inputs[i + 1].type(), NoneType):
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:3402]: what happens for using optional block inputs with loop opset < 15. This code looks like it will be silently not setting types for it. Will this end up with invalid model?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -525,6 +547,9 @@ def _model_to_graph(model, args, verbose=False,
 
     _set_input_and_output_names(graph, input_names, output_names)
     params_dict = _get_named_param_dict(graph, params)
+    # make sure that the param dict and the graph match each other
```
Comment (BowenBao) [inline torch/onnx/utils.py:550]: Do you know what's the rationale behind the changes around this block?

**Review by BowenBao — COMMENTED** [review]
Thank you @garymm for updating this pull request! I'm leaving some comments and thoughts as looking through the code.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -970,77 +958,108 @@ def forward(self, x, y=None, z=None):
...
+        # With kwargs dict.
+        # Requires input_names to be set so that we can feed the inputs properly into ORT.
+
+        # TODO: Export default values as ONNX initializers, then this should not raise.
+        # Default values are accessible via FunctionSchema.
+        with self.assertRaisesRegex(ValueError, "Model requires 3 inputs. Input Feed contains 2"):
+            self.run_test(model, (x, {"y": y}), input_names=("x", "y"))
+
+        # Setting any input to None in tracing means that
+        # input doesn't exist at all in the exported model, so input_names doesn't match.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:975]: todo: check what is the current warning/error message.

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 5]
```diff
@@ -190,14 +190,26 @@ std::vector<Value*> ConvertSequenceDependencies(Node* node, int opset_version) {
...
       if (output->node()->owningBlock() != block) {
-        Node* id_node = block->owningGraph()->create(onnx::Identity);
-        id_node->insertBefore(block->return_node());
-        id_node->addInput(output);
-        id_node->output()->copyMetadata(output);
-        id_node->copyMetadata(n);
-        block->return_node()->replaceInputWith(output, id_node->output());
+        if (auto none_output_type = output->type()->cast<NoneType>() &&
+                index + 1 < block->inputs().size()) {
+          auto t = block->inputs().at(index + 1)->type()->cast<OptionalType>();
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:198]: This seems to be covering cases where identity takes optional input, which should be covered by identity in opset 16.
todo:  try remove and evaluate if existing test case failed. To keep this at least a fix to increment index through looping is needed.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:975]: current error message: "RuntimeError: number of input names provided (3) exceeded number of inputs (2)"

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 6]
```diff
@@ -235,6 +247,20 @@ void FixupONNXLoopNodeInputs(Node* node) {
         InsertCastForCond(next_cond_val, graph, sub_block->return_node());
     cast_node->copyMetadata(node);
   }
+
+  size_t index = 0;
+  for (auto input : node->inputs()) {
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:252]: note: reactively convert aten None to onnx Optional, when they are inputs to loop.

**Review by garymm — COMMENTED** [review]
@BowenBao I think this is ready for review. The check failures seem unrelated.

LMK if you'd like to do this in real time together.

**Comment by BowenBao** [comment]
Thanks @garymm, I will try to read through the code changes first.

**Comment by BowenBao** [comment]
Conflicting files due to out of sync, @garymm could you please do another rebase with latest onnx_ms_1 branch?

**Comment by garymm** [comment]
@BowenBao rebased

File: test/onnx/test_pytorch_onnx_no_runtime.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        self.assertTrue(output_0_optional_type.elem_type.HasField("tensor_type"))
+        output_0_tensor_type = output_0_optional_type.elem_type.tensor_type
+        self.assertEqual(
+            output_0_tensor_type.elem_type,
+            symbolic_helper.scalar_type_to_onnx[
+                symbolic_helper.scalar_type_to_pytorch_type.index(x.dtype)])
+        self.assertEqual(len(output_0_tensor_type.shape.dim), len(x.shape))
+
+
+instantiate_parametrized_tests(TestOptionalOutput)
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_no_runtime.py:76]: Thanks for creating this test file. The goal of testing export without running exported model in backend seems to align with many test cases in test_utility_funs.py, and those test cases are more suited in this file. Should we move them over in next steps?

File: test/onnx/test_pytorch_onnx_no_runtime.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            model, (x,), f, opset_version=15,
+            # Ensure condition is not constant
+            dynamic_axes={"x": {0: "condition"}}, input_names=["x"])
+        exported = onnx.load_from_string(f.getvalue())
+        output_0_type = exported.graph.output[0].type
+        self.assertTrue(output_0_type.HasField("optional_type"))
+        output_0_optional_type = output_0_type.optional_type
+        self.assertTrue(output_0_optional_type.elem_type.HasField("tensor_type"))
+        output_0_tensor_type = output_0_optional_type.elem_type.tensor_type
+        self.assertEqual(
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_no_runtime.py:69]: This part seems very verbose. Can it be done by direct onnx TypeProto comparison, with hope that onnx TypeProto construction of optional_type with the expected element tensor type is straightforward and simple?

File: test/onnx/test_pytorch_onnx_no_runtime.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            output_0_tensor_type.elem_type,
+            symbolic_helper.scalar_type_to_onnx[
+                symbolic_helper.scalar_type_to_pytorch_type.index(x.dtype)])
+        self.assertEqual(len(output_0_tensor_type.shape.dim), len(x.shape))
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_no_runtime.py:73]: Should test cover what dimension size is? is it a symbol or empty?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 10]
```diff
@@ -109,17 +108,30 @@ def inline_flatten_list(inputs, res_list):
...
+    for input_name, input in kw_inputs.items():
+        ort_inputs[input_name] = to_numpy(input)
+    inputs = to_numpy(inputs)
+    ort_sess_inputs = ort_sess.get_inputs()
+    for i, input in enumerate(inputs):
+        input_name = ort_sess_inputs[i].name
+        if input_name in ort_inputs:
+            raise ValueError(
+                f"got too many positional inputs. inputs: {inputs}. kw_inputs: {kw_inputs}")
+        ort_inputs[input_name] = input
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:134]: are the above change mainly for covering extra dict input that weren't covered in previous prs, and unrelated to optional input?

edit: looking at other parts of code, looks like this is required for optional.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -824,23 +839,23 @@ def forward(self, x_in):
         x = {torch.tensor(1.): torch.randn(1, 2, 3)}
         self.run_test(MyModel(), (x, {}))
 
-    @disableScriptTest()
+    @skipScriptTest()
     def test_dict_str(self):
         class MyModel(torch.nn.Module):
-            def forward(self, x_in):
+            def forward(self, x_in: Dict[str, Tensor]) -> Dict[str, Tensor]:
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:845]: does this change affect exporting?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 12]
```diff
@@ -978,88 +963,124 @@ def forward(self, x, y=None, z=None):
...
-        self.run_test(AllOptionalModel(), (y, None), input_names=["input_y"])
-        self.run_test(AllOptionalModel(), (None, z), input_names=["input_z"])
-        # With optional arguments dictionary
-        self.run_test(AllOptionalModel(), {"y": y, "z": None}, input_names=["input_y"])
-        self.run_test(AllOptionalModel(), {"y": None, "z": z}, input_names=["input_z"])
-
-    def test_input_as_output(self):
+        x = torch.randn(2, 3)
+        model = Model()
+        self.run_test(model, (x, None))
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1003]: With tracing, do these tests capture `if` operator in exported graph?

I wonder if it'd work when providing extra set of inputs (e.g. (None, other_tensor)) to `run_test`.

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:1003]: from comments looks like the answer to above question is it won't capture if and it won't work with other inputs combination in tracing ... 

Is it reasonable to assume that for optional input feature to really be effective, it would require scripting?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 13]
```diff
@@ -978,88 +963,124 @@ def forward(self, x, y=None, z=None):
...
         class Model(torch.nn.Module):
-            def __init__(self):
-                super().__init__()
-
             def forward(self, x: int, y):
                 return x + y
 
         x = 3
-        y = torch.randint(10, (2, 3, 4))
+        y = torch.ones(2, 3, 4)
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1083]: any particular reason for `randint -> ones`?
edit: dtype of y becomes float instead of int with this change.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 14]
```diff
@@ -1732,12 +1753,13 @@ def forward(self, x: int, y: int):
         y = 2
         self.run_test(ArithmeticModule(), (x, y))
 
-    @disableScriptTest()
+    # In tracing, None outputs are removed. In scripting they're kept but
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1756]: does this mean we can enable test for optional outputs? e.g. None in one branch, some tensor in other.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 15]
```diff
@@ -8327,7 +8340,7 @@ def make_input(batch_size):
         other_input = make_input(RNN_BATCH_SIZE + 1)
         self.run_test(model, other_input, batch_size=RNN_BATCH_SIZE + 1)
 
-    @disableScriptTest()  # TODO: RuntimeError: Exporting the operator __is_ to ONNX is not supported
+    @skipScriptTest()  # TODO: https://msdata.visualstudio.com/Vienna/_workitems/edit/1253950
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:8343]: __is_ operator is already added. Is this still failing but for different reason?

**Review by BowenBao — COMMENTED** [review]
Thank you for taking over this PR. Still reading through changes, leaving questions/comments along the way

File: test/onnx/test_pytorch_onnx_no_runtime.py (hunk 7 above)
Comment (garymm) [reply test/onnx/test_pytorch_onnx_no_runtime.py:76]: Sure. I think we should also move over tests from test/jit/test_onnx_export.py. Tracked by https://msdata.visualstudio.com/Vienna/_workitems/edit/1593350.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 14 above)
Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:1756]: Yes. Tests for that are in test/onnx/test_pytorch_onnx_no_runtime.py

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 15 above)
Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:8345]: Yes, currently fails with:
```
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                                │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 8360, in test_transformer_encoder                                                                                                                                                                                                                                   │·········
    self.run_test(MyModule(512, 8, 2048 , 0., 3), (x,), atol=1e-6)                                                                                                                                                                                                                                                                                                │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 347, in run_test                                                                                                                                                                                                                                                    │·········
    _run_test(script_model, scripting_remained_onnx_input_idx, flatten=False)                                                                                                                                                                                                                                                                                     │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 327, in _run_test                                                                                                                                                                                                                                                   │·········
    return run_model_test(self, m, batch_size=batch_size,                                                                                                                                                                                                                                                                                                         │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 185, in run_model_test                                                                                                                                                                                                                                              │·········
    ort_sess = convert_to_onnx(model, input=input, opset_version=self.opset_version,                                                                                                                                                                                                                                                                              │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 88, in convert_to_onnx                                                                                                                                                                                                                                              │·········
    torch.onnx._export(model, input_copy, f,                                                                                                                                                                                                                                                                                                                      │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/__init__.py", line 50, in _export                                                                                                                                                                                                                                                                          │·········
    result = utils._export(*args, **kwargs)                                                                                                                                                                                                                                                                                                                       │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 751, in _export                                                                                                                                                                                                                                                                            │·········
    _model_to_graph(model, args, verbose, input_names,                                                                                                                                                                                                                                                                                                            │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 540, in _model_to_graph                                                                                                                                                                                                                                                                    │·········
    graph = _optimize_graph(graph, operator_export_type,                                                                                                                                                                                                                                                                                                          │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 229, in _optimize_graph                                                                                                                                                                                                                                                                    │·········
    graph = torch._C._jit_pass_onnx(graph, operator_export_type)                                                                                                                                                                                                                                                                                                  │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/__init__.py", line 368, in _run_symbolic_function                                                                                                                                                                                                                                                          │·········
    return utils._run_symbolic_function(*args, **kwargs)                                                                                                                                                                                                                                                                                                          │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 1104, in _run_symbolic_function                                                                                                                                                                                                                                                            │·········
    return symbolic_fn(g, *inputs, **attrs)                                                                                                                                                                                                                                                                                                                       │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_opset10.py", line 27, in div                                                                                                                                                                                                                                                                      │·········
    return torch.onnx.symbolic_opset9.true_divide(g, self, other)                                                                                                                                                                                                                                                                                                 │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_opset9.py", line 196, in true_divide                                                                                                                                                                                                                                                              │·········
    if sym_help._is_fp(self) or sym_help._is_fp(other):                                                                                                                                                                                                                                                                                                           │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_helper.py", line 340, in _is_fp                                                                                                                                                                                                                                                                   │·········
    type = value.type().scalarType()                                                                                                                                                                                                                                                                                                                              │·········
RuntimeError: rINTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h":540, please report a bug to PyTorch.
```

**Review by garymm — COMMENTED** [review]
Thanks for the review.

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 16]
```diff
@@ -504,9 +670,14 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
...
     case ::c10::onnx::Loop: {
+      // inputs (0, 1) are (i, cond).
       auto loop_carried_output_size = n->blocks().at(0)->inputs().size() - 2;
       for (auto i : c10::irange(n->outputs().size())) {
-        auto type = n->blocks().at(0)->outputs().at(i + 1)->type();
+        // Get the type from the input rather than the output
+        // to handle the case where a block input is Optional but the
+        // output is not (i.e. if the loop executes > 0 times, the
+        // output will not be None).
+        auto type = n->blocks().at(0)->inputs().at(i + 2)->type();
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:680]: Tests are failing at this line in the case where `ConvertSequenceDependencies` is triggered because that breaks the expectation that the block inputs contain all the block outputs. I'm looking into how to fix this. Let me know if you have ideas.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 17]
```diff
@@ -321,17 +333,20 @@ def _run_test(m, remained_onnx_input_idx, flatten=True):
...
+            tracing_remained_onnx_input_idx = remained_onnx_input_idx["tracing"]
         else:
             scripting_remained_onnx_input_idx = remained_onnx_input_idx
             tracing_remained_onnx_input_idx = remained_onnx_input_idx
 
-        if self.is_script_test_enabled and not isinstance(model, torch.jit.ScriptModule):
-            script_model = torch.jit.script(model)
+        is_script = isinstance(model, (torch.jit.ScriptModule, torch.jit.ScriptFunction))
+
+        if self.is_script_test_enabled:
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:344]: TODO: audit tests that have `@skipScriptTest` but use `torch.jit.script` or `torch.jit.trace` explicitly in test body.

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp (hunk 16 above)
Comment (garymm) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp]: Fixed!

**Review by garymm — COMMENTED** [review]
@BowenBao I fixed the failing tests. Hopefully everything is good to go now. Please take a look. I'd really like to get this merged next week if possible.

File: torch/onnx/utils.py [hunk 18]
```diff
@@ -401,11 +405,34 @@ def _get_param_count_list(method_graph, args_params):
...
     return param_count_list
 
 
+def _resolve_and_flatten_graph_inputs(model, args_params):
+    sig = _signature(model)
+    param_keys = list(sig.parameters.keys())
+    resolved_args = []  # type: ignore[var-annotated]
+    if isinstance(model, (torch.jit.ScriptFunction, torch.jit.ScriptModule)):
+        for i, var in enumerate(args_params):
+            if var is None:
```
Comment (BowenBao) [inline torch/onnx/utils.py:421]: I feel the purpose is to grab default value when argument `var` is omitted in example inputs. But how does it work? I thought `var is None` is supposed to mean `None` passed as input explicitly. Let me know where I'm wrong.

File: torch/onnx/utils.py [hunk 19]
```diff
@@ -401,11 +405,34 @@ def _get_param_count_list(method_graph, args_params):
...
+                resolved_args.append(sig.parameters[param_keys[i]].default)
+            elif isinstance(var, tuple):
```
Comment (BowenBao) [inline torch/onnx/utils.py:423]: nit: not sure if nested tuple is something we'd care about.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 12 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:1003]: Thanks for explanation. Let's keep it to ensure tracing route is compatible with this signature and None/Optional usage.

File: torch/onnx/utils.py (hunk 19 above)
Comment (garymm) [reply torch/onnx/utils.py]: Moot now that this function is gone.

**Review by garymm — COMMENTED** [review]
Thanks for the review!

File: test/onnx/test_pytorch_onnx_no_runtime.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
+# Owner(s): ["module: onnx"]
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_no_runtime.py:1]: Add to `scripts/onnx/test.sh` if intend to run in CI.

**Comment by garymm** [comment]
Converting to draft until I have time to understand the shape and type inference issues better and add more tests.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 21]
```diff
@@ -1828,6 +1823,14 @@ void SpecialPostProcess(Node* n) {
       }
       break;
     }
+    case ::c10::onnx::Optional: {
```
[the remark this answers is not in the document]
Comment (garymm) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp]: Deleted. It seems it may not be needed.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 22]
```diff
@@ -237,6 +241,12 @@ bool IsValidONNXNode(const Node* n) {
     }
   }
 
+  for (auto inp : n->inputs()) {
+    if (inp->type() == NoneType::get()) {
+      return false;
```
[the remark this answers is not in the document]
Comment (garymm) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:231]: Found and fixed the only known case in `FixupONNXSubblockOutputs` by using empty `onnx::Optional` instead of `onnx::Identity(None)`. Will address other cases as needed.

**Review by garymm — COMMENTED** [review]
If CI passes I think this might be ready to go!

**Review by BowenBao — COMMENTED** [review]
Some test failures from `test_pytorch_onnx_caffe2.py`. It looks having to do with how input arguments are parsed.

**Review by BowenBao — APPROVED** [review]
Thanks again for completing this PR!

**Review by garymm — COMMENTED** [review]
Fixed the last remaining issues. Going to merge now.
